Kernel/Graphics/VirtIOGPU/GraphicsAdapter.h

```diff
@@ -6,28 +6,86 @@
 
 #pragma once
 
+#include <AK/BinaryBufferWriter.h>
+#include <AK/DistinctNumeric.h>
+#include <Kernel/Bus/VirtIO/Device.h>
+#include <Kernel/Bus/VirtIO/Queue.h>
+#include <Kernel/Devices/BlockDevice.h>
+#include <Kernel/Graphics/GenericGraphicsAdapter.h>
 #include <Kernel/Graphics/VirtIOGPU/Console.h>
-#include <Kernel/Graphics/VirtIOGPU/FrameBufferDevice.h>
-#include <Kernel/Graphics/VirtIOGPU/GPU.h>
+#include <Kernel/Graphics/VirtIOGPU/FramebufferDevice.h>
+#include <Kernel/Graphics/VirtIOGPU/Protocol.h>
 
 namespace Kernel::Graphics::VirtIOGPU {
 
+#define VIRTIO_GPU_F_VIRGL (1 << 0)
+#define VIRTIO_GPU_F_EDID (1 << 1)
+
+#define VIRTIO_GPU_FLAG_FENCE (1 << 0)
+
+#define CONTROLQ 0
+#define CURSORQ 1
+
+#define MAX_VIRTIOGPU_RESOLUTION_WIDTH 3840
+#define MAX_VIRTIOGPU_RESOLUTION_HEIGHT 2160
+
+#define VIRTIO_GPU_EVENT_DISPLAY (1 << 0)
+
+class FramebufferDevice;
 class GraphicsAdapter final
-    : public GraphicsDevice
-    , public PCI::Device {
+    : public GenericGraphicsAdapter
+    , public VirtIO::Device {
     AK_MAKE_ETERNAL
+    friend class FramebufferDevice;
 
 public:
     static NonnullRefPtr<GraphicsAdapter> initialize(PCI::DeviceIdentifier const&);
 
     virtual bool framebuffer_devices_initialized() const override { return m_created_framebuffer_devices; }
 
+    // FIXME: There's a VirtIO VGA GPU variant, so we should consider that
+    virtual bool vga_compatible() const override { return false; }
+
+    virtual void initialize() override;
```

Comment:
Nit: At some point, this should return a result, instead of void;

Reply:
Working on this in another pull request (#9818).